### PR TITLE
Optimize SQL statements and transactions for cacheless profile

### LINKED ISSUE #49697: Avoid separate transactions and extra SQL statements in cacheless mode
### Description

The tests are green, still they can be optimized: 

* For password credential grant, there is a temporary authentication session that is deleted at the end of the transaction. This should not touch the database
* When JPA based authentication sessions are used, this uses pessimistic locking, and due to that the inserts for user and client sessions can not conflict. 
* For user and client sessions, one can cache that the session does not exist, avoiding an extra select

### PR DESCRIPTION
**Pull request by ahus1**
Closes #49697

Before when submitting a username and password to the login form: 

603032181-4f02e827-9f1d-4c91-bcf6-068651158d13

After: 

image

Before for resource owner credential grant: 

image

After with fewer spans, no insert/update/delete for the authentication session, and no separate transaction for the user session: 

image

### CODE REVIEW

**Comment by ahus1** [comment]
@pruivo - FYI. Local smoke-tests look promising. Let's see what the CI says about this.

**Comment by pruivo** [comment]
@ahus1 Maybe because it is late, but aren't those images the same? Don't they have the same number of insers/update/deletes?

**Comment by ahus1** [comment]
It seems that I was lazy to add coloring: 

* There are two selects now removed that are colored in yellow in the now updated screenshot
* Previously there were two commits, and now there is one (the one marked in yellow is now inlined)

As we learned that commits to a transaction that is writing are expensive, the second one should have the bigger impact of the two.

**Comment by ahus1** [comment]
I can add another example for password credential grants later but not today. With that example, you would see the insert and delete auf the root authentication session have been eliminated after the change,

**Comment by ahus1** [comment]
Added another screenshot for the resource owner credential grant. 

@ruchikajha95 - it would be great if you could give it a review next week. Thanks!

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/AuthenticationSessionEntity.java [hunk 1]
```diff
@@ -223,4 +226,33 @@ public String toString() {
                 ", rootAuthenticationSessionId=" + (rootAuthenticationSession != null ? rootAuthenticationSession.getId() : null) + // avoid lazy-load just for a log message.
                 '}';
     }
+
+    public static class Key implements Serializable {
```
Comment (pruivo) [inline model/jpa/src/main/java/org/keycloak/authentication/jpa/AuthenticationSessionEntity.java]: can be a record outside the class, like 
https://github.com/keycloak/keycloak/blob/5274246ade2f0a4731ee5004cf29eec4ee58b4aa/model/jpa/src/main/java/org/keycloak/loginfailures/jpa/LoginFailureKey.java#L23

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java [hunk 2]
```diff
@@ -50,10 +56,30 @@ public JpaAuthenticationSessionProvider(KeycloakSession session, int authSession
...
         var model = RootAuthenticationSessionAdapter.create(session, realm, SecretGenerator.SECURE_ID_GENERATOR.get(), Time.currentTime(), authSessionsLimit);
-        getEntityManager().persist(model.getEntity());
+        // Those newly created authentication sessions with a random ID do not exist in the database, so there can not be any conflict.
+        // For resource owner password grants, those sessions are created temporarily, so we only insert them if they are not removed within the same session.
+        transientSessions.put(model.getEntity().getId(), model);
+        prepareTransaction();
         return model;
     }
 
+    private void prepareTransaction() {
```
Comment (pruivo) [inline model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java:67]: This looks quite ugly to me (personal opinion). 
I would make `JpaAuthenticationSessionProvider` extend `AbstractKeycloakTransaction` and register the prepare in the factory.

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProviderFactory.java [hunk 3]
```diff
@@ -38,11 +38,12 @@
 import org.keycloak.provider.ProviderConfigProperty;
 import org.keycloak.provider.ProviderConfigurationBuilder;
 import org.keycloak.provider.ServerInfoAwareProviderFactory;
+import org.keycloak.sessions.AuthenticationSessionProvider;
 import org.keycloak.sessions.AuthenticationSessionProviderFactory;
 
 import org.jboss.logging.Logger;
 
-public class JpaAuthenticationSessionProviderFactory implements AuthenticationSessionProviderFactory<JpaAuthenticationSessionProvider>, EnvironmentDependentProviderFactory, ServerInfoAwareProviderFactory {
+public class JpaAuthenticationSessionProviderFactory implements AuthenticationSessionProviderFactory<AuthenticationSessionProvider>, EnvironmentDependentProviderFactory, ServerInfoAwareProviderFactory {
```
Comment (pruivo) [inline model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProviderFactory.java:46]: Why is this change required?

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/RootAuthenticationSessionAdapter.java [hunk 4]
```diff
@@ -134,7 +134,7 @@ public AuthenticationSessionModel createAuthenticationSession(ClientModel client
 
     @Override
     public void removeAuthenticationSessionByTabId(String tabId) {
-        entity.getAuthenticationSessions().remove(tabId);
+        session.getProvider(JpaConnectionProvider.class).getEntityManager().remove(entity.getAuthenticationSessions().remove(tabId));
```
Comment (pruivo) [inline model/jpa/src/main/java/org/keycloak/authentication/jpa/RootAuthenticationSessionAdapter.java:137]: Doesn't hibernate remove the entity when it is removed from the map?

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java [hunk 5]
```diff
@@ -82,6 +108,10 @@ public RootAuthenticationSessionModel getRootAuthenticationSession(RealmModel re
         if (id == null) {
             return null;
         }
+        var model = transientSessions.get(id);
+        if (model != null && Objects.equals(model.getRealm().getId(), realm.getId())) {
```
Comment (pruivo) [inline model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java:103]: What's the scenario where the realms do not match? I see this pattern in the code, but I haven't seen a comment explaining why it can happen.

**Review by pruivo — DISMISSED** [review]

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java (hunk 5 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java:103]: Maybe there has been a mixup / CVE in the past with cross-realm handling, and now we're doing all the time. After all, the realm ID is not part of the primary key. Not sure how one can force that from an attacker side ... consider it a second line of defense for now.

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/RootAuthenticationSessionAdapter.java (hunk 4 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/authentication/jpa/RootAuthenticationSessionAdapter.java:137]: When you look the ROCG example above (third screenshot), you'll see that the DELETE of of the authentication session was actually missing. This was then fixed by removing the persist cascade form the child table. At some point Hibernate then complained about an unmanaged entity of the root authentication session, that then vanished once I did this one.

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProviderFactory.java (hunk 3 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProviderFactory.java:46]: When I added this `instanceof` check, the JVM complained that this would not work without any casting.

https://github.com/keycloak/keycloak/blob/b80ede179470305ca870e461322d0e2fa113e268/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java#L212-L214

Once I change this, it started to compile.

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java (hunk 2 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java:67]: changed in next commit. If any other changes are necessary, let's do them in a follow-up.

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/AuthenticationSessionEntity.java (hunk 1 above)
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/authentication/jpa/AuthenticationSessionEntity.java]: done, commit pushed

File: model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java (hunk 5 above)
Comment (ruchikajha95) [reply model/jpa/src/main/java/org/keycloak/authentication/jpa/JpaAuthenticationSessionProvider.java:103]: Should we rather  a brief comment or // NOTE: above this line just to capture that context? It'll be helpful in future to understand why the realm check is here even though it's not part of the PK.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java [hunk 6]
```diff
@@ -82,6 +83,21 @@ public void testConcurrentNotesChange() throws InterruptedException {
...
+                        RealmModel realm = session.realms().getRealm(realmId);
+                        session.getContext().setRealm(realm);
+                        ClientModel client = realm.getClientByClientId("client" + n);
+
+                        UserSessionModel uSession = session.sessions().getUserSession(realm, uId);
+                        session.sessions().createClientSession(realm, client, uSession);
+
+                        return null;
+                    }));
+        }
```
Comment (ruchikajha95) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java:100]: this covers the locking path .Should we also add a test asserting zero SQL writes on the create-then-remove case?

**Comment by ahus1** [comment]
@ruchikajha95 - be build is now green. I needed to do a small update as the test was unstable as the number of DB operations differed when running the single test case vs running the whole test file. This was due to the browser still having some cookies. Once I removed those, and reset the browser to the first page, it was then stable. 

Please give it one more review. Thanks!

**Review by ruchikajha95 — APPROVED** [review]
LGTM

**Review by pedroigor — APPROVED** [review]
